### PR TITLE
Bug 2039345: Verify against mininimal IPv6 MTU value for clusters with IPv6 networks

### PR DESCRIPTION
**Pull request by kyrtapz**
Additionally added verification of migration MTU values.

/cc @ricky-rav @jcaamano 

Signed-off-by: Patryk Diak <pdiak@redhat.com>

### CODE REVIEW

**Review by flavio-fernandes — COMMENTED** [review]
/lgtm

**Comment by andreaskaris** [comment]
/lgtm 

IMO nice to have: It'd be super nice to refer to RFC8200 (IPv6) and RFC791 (IPv4) in a comment for reference

**Comment by jcaamano** [comment]
I talked with @kyrtapz about applying these validations against migration MTUs as well.

File: pkg/network/ovn_kubernetes_test.go [hunk 1]
```diff
@@ -669,11 +677,37 @@ func TestOVNKubernetesIsSafe(t *testing.T) {
 
 	next.Migration.MTU.Network.From = prev.DefaultNetwork.OVNKubernetesConfig.MTU
 
+	// invalid Migration.MTU.Host.To, lower than minimum MTU for IPv4
```
Comment (andreaskaris) [inline pkg/network/ovn_kubernetes_test.go:680]: Shouldn't the comment be: `// invalid Migration.MTU.Network.To, lower than minimum MTU for IPv4`

File: pkg/network/ovn_kubernetes_test.go [hunk 2]
```diff
@@ -669,11 +677,37 @@ func TestOVNKubernetesIsSafe(t *testing.T) {
...
+	next.Migration.MTU.Network.To = ptrToUint32(100)
+	errs = isOVNKubernetesChangeSafe(prev, next)
+	g.Expect(errs).To(HaveLen(1))
+	g.Expect(errs[0]).To(MatchError(fmt.Sprintf("invalid Migration.MTU.Network.To(%d), has to be in range: %d-%d", *next.Migration.MTU.Network.To, MinMTUIPv4, MaxMTU)))
+
+	// invalid Migration.MTU.Host.To, higher than maximum MTU for IPv4
```
Comment (andreaskaris) [inline pkg/network/ovn_kubernetes_test.go:686]: Shouldn't the comment be: `// invalid Migration.MTU.Network.To, higher than maximum MTU for IPv4`

File: pkg/network/ovn_kubernetes_test.go [hunk 3]
```diff
@@ -669,11 +677,37 @@ func TestOVNKubernetesIsSafe(t *testing.T) {
...
+	next.Migration.MTU.Network.To = ptrToUint32(65537)
+	errs = isOVNKubernetesChangeSafe(prev, next)
+	g.Expect(errs).To(HaveLen(2))
+	g.Expect(errs[0]).To(MatchError(fmt.Sprintf("invalid Migration.MTU.Network.To(%d), has to be in range: %d-%d", *next.Migration.MTU.Network.To, MinMTUIPv4, MaxMTU)))
+
+	next.Migration.MTU.Network.To = ptrToUint32(1300)
+
```
Comment (andreaskaris) [inline pkg/network/ovn_kubernetes_test.go:693]: Do we need tests for `next.Migration.MTU.Machine.To` as well?

File: pkg/network/openshift_sdn.go [hunk 4]
```diff
@@ -204,6 +204,13 @@ func isOpenShiftSDNChangeSafe(prev, next *operv1.NetworkSpec) []error {
 			if checkPrevMTU && *next.Migration.MTU.Network.From != *pn.MTU {
 				errs = append(errs, errors.Errorf("invalid Migration.MTU.Network.From(%d) not equal to the currently applied MTU(%d)", *next.Migration.MTU.Network.From, *pn.MTU))
 			}
+
+			if *next.Migration.MTU.Network.To < MinMTUIPv4 || *next.Migration.MTU.Network.To > MaxMTU {
+				errs = append(errs, errors.Errorf("invalid Migration.MTU.Network.To(%d), has to be in range: %d-%d", *next.Migration.MTU.Network.To, MinMTUIPv4, MaxMTU))
```
Comment (andreaskaris) [inline pkg/network/openshift_sdn.go:209]: Can you add tests to `TestOpenShiftSDNIsSafe` in `openshift_sdn_test.go`?

File: pkg/network/ovn_kubernetes_test.go (hunk 3 above)
Comment (andreaskaris) [inline pkg/network/ovn_kubernetes_test.go:693]: Do we need tests for `next.Migration.MTU.Machine.To` as well? (at least for max)

**Review by andreaskaris — CHANGES_REQUESTED** [review]

File: pkg/network/ovn_kubernetes_test.go (hunk 3 above)
Comment (kyrtapz) [reply pkg/network/ovn_kubernetes_test.go:693]: Yeah, added that.

File: pkg/network/openshift_sdn.go (hunk 4 above)
Comment (andreaskaris) [inline pkg/network/openshift_sdn.go:209]: I just saw this now when re-reviewing the changes and having it force fail:
~~~
[akaris@linux network (pull1276)]$ go test -v -run "TestOVNKubernetesIsSafe" .
=== RUN   TestOVNKubernetesIsSafe
2022/01/25 14:28:47 Detected uplink MTU 1500
2022/01/25 14:28:47 Detected uplink MTU 1500
    ovn_kubernetes_test.go:717: 
        Expected
            <*errors.fundamental | 0xc0004e15f0>: {
                msg: "invalid Migration.MTU.Machine.To(65636), has to be in range: 1280-65536",
                stack: [0x12fb10e, 0x1319c5c, 0x50d182, 0x46a381],
            }
        to match error
            <string>: iinvalid Migration.MTU.Machine.To(65636), has to be in range: 1280-65536
--- FAIL: TestOVNKubernetesIsSafe (0.00s)
FAIL
FAIL	github.com/openshift/cluster-network-operator/pkg/network	0.014s
FAIL
~~~

All of these error messages here:
~~~
errs = append(errs, errors.Errorf("invalid Migration.MTU.Network.To(%d), has to be in range: %d-%d", *next.Migration.MTU.Network.To, MinMTUIPv4, MaxMTU))
~~~

Should have `MaxMTU - 1`, correct? 
~~~
errs = append(errs, errors.Errorf("invalid Migration.MTU.Network.To(%d), has to be in range: %d-%d", *next.Migration.MTU.Network.To, MinMTUIPv4, MaxMTU-1))
~~~

And then, we'd get: `invalid Migration.MTU.Machine.To(65636), has to be in range: 1280-65535`

Sorry for not having seen that earlier.

File: pkg/network/openshift_sdn_test.go [hunk 5]
```diff
@@ -367,11 +367,32 @@ func TestOpenShiftSDNIsSafe(t *testing.T) {
...
+
+	next.Migration.MTU.Network.To = ptrToUint32(1300)
+
 	// invalid Migration.MTU.Host.To, not big enough to accommodate next.Migration.MTU.Network.To with encap overhead
 	next.Migration.MTU.Network.To = ptrToUint32(1500)
 	errs = isOpenShiftSDNChangeSafe(prev, next)
 	g.Expect(errs).To(HaveLen(1))
 	g.Expect(errs[0]).To(MatchError(fmt.Sprintf("invalid Migration.MTU.Machine.To(%d), has to be at least %d", *next.Migration.MTU.Machine.To, *next.Migration.MTU.Network.To+50)))
+
+	// invalid Migration.MTU.Machine.To, higher that max MTU
```
Comment (andreaskaris) [inline pkg/network/openshift_sdn_test.go:390]: higher than

File: pkg/network/ovn_kubernetes_test.go [hunk 6]
```diff
@@ -669,11 +677,44 @@ func TestOVNKubernetesIsSafe(t *testing.T) {
...
+		{
+			CIDR:       "fd00:1:2:3::/64",
+			HostPrefix: 56,
+		},
+	}
+	errs = isOVNKubernetesChangeSafe(prev, next)
+	g.Expect(errs).To(HaveLen(1))
+	g.Expect(errs[0]).To(MatchError(fmt.Sprintf("invalid Migration.MTU.Network.To(%d), has to be in range: %d-%d", *next.Migration.MTU.Network.To, MinMTUIPv6, MaxMTU)))
+
+	// invalid Migration.MTU.Machine.To, higher that max MTU
```
Comment (andreaskaris) [inline pkg/network/ovn_kubernetes_test.go:712]: higher than

**Review by andreaskaris — CHANGES_REQUESTED** [review]

**Pull request merged**